### PR TITLE
add a experiment API for upsert heap memory estimation

### PR DESCRIPTION
**Pull request by deemoliu**
## Description

Added a new rest resource and an endpoint for estimating heap memory usage for Pinot upsert table.
Related Pinot issue: #6729

## Release Notes

Added a http endpoint for estimate the heap usage for Pinot upsert table  per partition based on PK cardinality.
- POST /upsert/estimateHeapUsage

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableExperimentRestletResource.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Api(tags = Constants.EXPERIMENT_TAG)
+@Path("/")
+public class PinotTableExperimentRestletResource {
```
Comment (chenboat) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableExperimentRestletResource.java:45]: This is resource is about UpsertTable only. Why not name it explicitly so that it is clear that this is about UpsertTable resource estimation?

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/ColumnStats.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+/*
+ * Container object for stats of Pinot columns, output by kafka sampler.
+ */
+public class ColumnStats {
+
+  public static final String CARDINALITY = "cardinality";
+  public static final String PRIMARY_KEY_SIZE = "primaryKeySize";
+  public static final String COMPARISON_COL_SIZE = "comparisonColSize";
+  public static final String PARTITIONS_NUM = "partitionsNums";
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/ColumnStats.java:34]:
```suggestion
  public static final String NUM_PARTITIONS = "numPartitions";
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+   * -comparisonColSize, it uses for calculating BytesPerValue
+   * ```
+   */
+  @POST
+  @Path("/heapUsage")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Estimate memory usage for an upsert table", notes =
+      "This API returns the estimated heap usage based on primary key column stats."
+          + " This allows us to estimate table size before onboarding.")
+  public String estimateHeapUsage(@QueryParam("tableConfig") String tableConfigStr,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:77]: This API is built as a tool instead of a regular restlet API as it doesn't rely on the information stored in the cluster. We should at least read the instance config info, calculate the partitions on each server, and estimate the heap usage per server

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    ObjectNode resultData = JsonUtils.newObjectNode();
+    List<String> primaryKeys = schema.getPrimaryKeyColumns();
+    // Estimated key space, it contains primary key columns
+    for (String primaryKey : primaryKeys) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
+      if (dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for primary key columns";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else if (dt == FieldSpec.DataType.STRING) {
+        bytesPerKey += columnStats.getPrimaryKeySize();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:112]: If primary key size is provided, we should skip this calculation and directly using the provided value

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      } else {
+        bytesPerKey += dt.size();
+      }
+    }
+    // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)>
+    int bytesPerValue =
+        tableConfig.getUpsertConfig().getComparisonColumn() != null ? 52 + columnStats.getComparisonColSize() : 64;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:119]: If comparison column is fixed width, no need to provide it in `columnStats`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    // Estimated key space, it contains primary key columns
+    for (String primaryKey : primaryKeys) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
+      if (dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for primary key columns";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else if (dt == FieldSpec.DataType.STRING) {
+        bytesPerKey += columnStats.getPrimaryKeySize();
+      } else {
+        bytesPerKey += dt.size();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:114]: We should probably also count the array size?

Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: gotcha, thanks Jackie. Java has a 24 bytes array overhead and there's also 8 bytes for the actual array object, so i added 32 bytes in bytes per key.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 5 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:119]: gotcha, thanks Jackie. I will infer the comparison size from Datatype. 
`    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
    if (comparisonColumn != null) {
      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
      if (dt == FieldSpec.DataType.STRING || dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST
          || dt == FieldSpec.DataType.MAP) {
        String msg = "Not support data types for the comparison column";
        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
      } else {
        bytesPerValue = 52 + columnStats.getComparisonColSize();
      }
    }`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 4 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: gotcha, updated.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 3 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: gotcha, I added  numPartitions, replicasPerPartition and memoryPerHost

File: pinot-controller/src/test/java/org/apache/pinot/controller/ControllerTestUtils.java [hunk 7]
```diff
@@ -471,6 +472,12 @@ public static Schema createDummySchema(String tableName) {
     return schema;
   }
 
+  public static Schema createDummySchemaForUpsertTable(String tableName) {
+    Schema schema = createDummySchema(tableName);
+    schema.setPrimaryKeyColumns(Lists.newArrayList("dimA"));
```
Comment (Jackie-Jiang) [inline pinot-controller/src/test/java/org/apache/pinot/controller/ControllerTestUtils.java:477]: (minor)
```suggestion
    schema.setPrimaryKeyColumns(Collections.singletonList("dimA"));
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+   * For columns stats, we need to gather the following stats
+   * ```
+   * -cardinality, a required information unique combination of primary keys.
+   * -primaryKeySize, it uses for calculating BytesPerKey.
+   * -comparisonColSize, it uses for calculating BytesPerValue.
+   * -partitionNums(optional), it uses for host assignment calculation.
+   * ```
+   */
+  @POST
+  @Path("/heapUsage")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:74]:
```suggestion
  @Path("/estimateHeapUsage")
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Estimate memory usage for an upsert table", notes =
+      "This API returns the estimated heap usage based on primary key column stats."
+          + " This allows us to estimate table size before onboarding.")
+  public String estimateHeapUsage(String tableSchemaConfigStr,
+      @ApiParam(value = "cardinality in string format", required = true) @QueryParam("cardinality") String cardinality,
+      @ApiParam(value = "primaryKeySize in string format") @QueryParam("primaryKeySize") String primaryKeySize,
+      @ApiParam(value = "numPartitions in string format") @QueryParam("numPartitions") String numPartitionsStr) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:83]: Both of them should be `int`, and we may put a default value `-1`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+    // Estimated key space, it contains primary key columns
+    int bytesPerKey = 0;
+    List<String> primaryKeys = schema.getPrimaryKeyColumns();
+
+    if (primaryKeySize != null) {
+      bytesPerKey += Integer.valueOf(primaryKeySize);
+    } else {
+      for (String primaryKey : primaryKeys) {
+        FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:108]: You may check `isFixedWidth()` and throw exception if it is not fixed size

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      // Java has a 24 bytes array overhead and there's also 8 bytes for the actual array object
+      bytesPerKey += 32;
+    }
+
+    // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)> and overhead.
+    int bytesPerValue = 64;
+    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
+    if (comparisonColumn != null) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
+      if (dt == FieldSpec.DataType.STRING || dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:128]: Same here, check if data type `isFixedWidth()`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+          || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for the comparison column";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else {
+        bytesPerValue = 52 + dt.size();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:133]: Is this correct? When comparison column is long, this does not match the timestamp one of 64 bytes

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+      }
+    }
+
+    resultData.put("bytesPerKey", bytesPerKey);
+    resultData.put("bytesPerValue", bytesPerValue);
+
+    long primaryKeyCardinality = Long.valueOf(cardinality);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:140]:
```suggestion
    long primaryKeyCardinality = Long.parseLong(cardinality);
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+   */
+  @POST
+  @Path("/heapUsage")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Estimate memory usage for an upsert table", notes =
+      "This API returns the estimated heap usage based on primary key column stats."
+          + " This allows us to estimate table size before onboarding.")
+  public String estimateHeapUsage(String tableSchemaConfigStr,
+      @ApiParam(value = "cardinality in string format", required = true) @QueryParam("cardinality") String cardinality,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:81]:
```suggestion
      @ApiParam(value = "cardinality in string format", required = true) @QueryParam("cardinality") long cardinality,
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+    // Use Partitions, replicas to calculate memoryPerHost for host assignment.
+    if (numPartitionsStr != null) {
+      int numPartitions = Integer.valueOf(numPartitionsStr);
+      if (numPartitions > 0) {
+        int replicasPerPartition = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+        double memoryPerHost = (totalSpace * replicasPerPartition * 1.0) / numPartitions;
+        resultData.put("numPartitions", numPartitions);
+        resultData.put("replicasPerPartition", replicasPerPartition);
+        resultData.put("memoryPerHost", memoryPerHost);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:157]: `totalSpacePerPartition(bytes)`?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+      }
+    }
+
+    resultData.put("bytesPerKey", bytesPerKey);
+    resultData.put("bytesPerValue", bytesPerValue);
+
+    long primaryKeyCardinality = Long.valueOf(cardinality);
+    long totalKeySpace = bytesPerKey * primaryKeyCardinality;
+    long totalValueSpace = bytesPerValue * primaryKeyCardinality;
+    long totalSpace = totalKeySpace + totalValueSpace;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:143]: (optional) Here we only calculate the map content size. Should we also consider the map entry size and the array size within the map?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/Constants.java [hunk 17]
```diff
@@ -46,6 +46,7 @@ private Constants() {
   public static final String ZOOKEEPER = "Zookeeper";
   public static final String APP_CONFIGS = "AppConfigs";
   public static final String PERIODIC_TASK_TAG = "PeriodicTask";
+  public static final String UPSERT_RESOURCE_TAG = "UpsertResource";
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/Constants.java:49]:
```suggestion
  public static final String UPSERT_RESOURCE_TAG = "Upsert";
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java [hunk 18]
```diff
@@ -194,7 +194,7 @@ public String forTableRebalance(String tableName, String tableType, boolean dryR
   }
 
   public String forTableReload(String tableName, TableType tableType, boolean forceDownload) {
-    String query = String.format("reload?type=%s&forceDownload=%s", tableType.name(), forceDownload);
+    String query = String.format("reload?forceDownload=%s&type=%s", forceDownload, tableType.name());
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java:197]: Please revert the unrelated changes. I think something is not properly merged

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 12 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:133]: thanks @Jackie-Jiang 

For values does it contains offset?
According to discussion in https://github.com/apache/pinot/issues/6729, we need to store RecordLocation, it contains <segmentName, DocId, Timestamp>, so maybe another <48+ 4 + 8>=64 bytes on it.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java (hunk 18 above)
Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java:197]: thanks for pointing out. reverted.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/Constants.java (hunk 17 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/Constants.java:49]: gotcha, updated.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 16 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:143]: The object overhead will vary between different machines. 

I found this https://dzone.com/articles/size-entry-map shows 81.4 bytes per concurrentHashmap entry, do you think this is accurate and want me to add that?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 15 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:157]: gotcha, thanks for suggestion.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+    resultData.put("totalSpace(bytes)", totalSpace);
+
+    // Use Partitions, replicas to calculate memoryPerHost for host assignment.
+    if (numPartitionsStr != null) {
+      int numPartitions = Integer.valueOf(numPartitionsStr);
+      if (numPartitions > 0) {
+        int replicasPerPartition = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+        double memoryPerHost = (totalSpace * replicasPerPartition * 1.0) / numPartitions;
+        resultData.put("numPartitions", numPartitions);
+        resultData.put("replicasPerPartition", replicasPerPartition);
```
[the remark this answers is not in the document]
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:156]: gotcha.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 14 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:81]: gotcha, updated.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 11 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java:128]: gotcha, thanks for suggestion!

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Api(tags = Constants.UPSERT_RESOURCE_TAG)
+@Path("/")
+public class PinotUpsertCapacityEstimationRestletResource {
```
[the remark this answers is not in the document]
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: thanks @Jackie-Jiang

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 12 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: `48 + 4 + 8 = 60 != 64` lol

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 16 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: In order to make more accurate estimation, we should count the object overhead because as you can see, the object overhead can be larger than the actual content within the entry.
I don't know the formula of object overhead from map size, so we need to do some research here. If you find that is too hard, we may add a TODO and address that later. But users should know this size does not include the overhead, so it can be quite off.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 12 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: gotcha, thanks. i will update the valueSpace with default timestamp from 64 bytes to 60 bytes.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java (hunk 16 above)
Comment (deemoliu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java]: Thanks for suggestion! I can add a Todo and address later. I will do some benchmark with heap dump.

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertRestletResource.java [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+   * For columns stats, we need to gather the following stats
+   * ```
+   * -cardinality, a required information unique combination of primary keys.
+   * -primaryKeySize, it uses for calculating BytesPerKey.
+   * -comparisonColSize, it uses for calculating BytesPerValue.
+   * -partitionNums(optional), it uses for host assignment calculation.
+   * ```
+   */
+  @POST
+  @Path("/estimateHeapUsage")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertRestletResource.java:74]:
```suggestion
  @Path("/upsert/estimateHeapUsage")
```

**Review by Jackie-Jiang — COMMENTED** [review]
Please update the release note of the PR description for the new added rest endpoint

**Comment by deemoliu** [comment]
> Please update the release note of the PR description for the new added rest endpoint

Thanks @Jackie-Jiang , done.
